### ISSUE TITLE
Multiple notebook to one kernel?

### ISSUE THREAD

**Issue by cqcn1991** [jupyter/notebook#1699]
I'm doing some complicated works with Jupyter Notebook, so the notebook is very long (https://github.com/cqcn1991/Wind-Speed-Analysis).

Sometimes, during the middle of the notebook, I want to do some additional analysis. Adding them directly into the current notebook may make it more complex, and breaks its current structure. I think it would be amazing if I can simply open another notebook, connect it to the existing notebook's kernel, then to do the additional analysis.

Maybe something like

    # In the new notebook
    connect_to 'exisiting_notebook_name' # get access to the existing notebook
    df.describe()
    # ...
    # some additional analysis works

I have asked it [here](http://stackoverflow.com/questions/39055213/jupyter-notebook-multiple-notebook-to-one-kernel), but didn't get any response. so I post here again to get some idea.

**Comment by takluyver** [comment]
You may be able to hack something together, but I think its's something we've considered before and aren't interested in supporting; the mechanisms to see and change which kernel a notebook is connected to would be a significant extra piece of user-facing complexity.

**Comment by cqcn1991** [comment]
@takluyver Then if there is any clue on this? I have no idea how to start to hack this.

**Comment by takluyver** [comment]
[Sessionmanager](https://github.com/jupyter/notebook/blob/master/notebook/services/sessions/sessionmanager.py) is what maintains the mapping between notebooks and kernels. I haven't thought at all about how you'd adapt it to do what you want, though.

**Comment by gshiba** [comment]
Not a separate notebook, and one cell only, but this extension may be useful: https://github.com/minrk/nbextension-scratchpad

**Comment by foice** [comment]
Let me add that splitting one's work on several notebooks helps to locate things end edit them. I think is the same basic principle why in any complex progam (e.g. in C) one uses many files to define stuff. I know that I can push code into some .py and import it, but what if I need to see graphics? (lots of graphics)

Furthermore by splitting notebook on several browser tabs might solve issues to java-script timeouts such as those of [this issue](https://github.com/jupyter/notebook/issues/2075). 

I am trying to transition to jupyter from Wolfram Mathematica (or similar closed source program) and it's quite a show stopper to be limited to one-notebook work.

**Comment by cqcn1991** [comment]
@foice same thinking, and I think it's not very difficult to implement this feature.

**Comment by ran-z** [comment]
I agree, this feature could be very helpful at keeping different aspects of analysis or research separate and compartmentalized.

I'm thinking perhaps another approach which could achieve a similar outcome but will be easier to implement, is to create an extension which allows to define regions in the notebook - and then you could use one cell to start a region and give it a name, and another cell to end the region, and the extension can allow region collapse, execution, etc.

**Comment by foice** [comment]
ok, I am not alone then! even the fact that cut and paste of cells works only within the same notebook is a bit of a limitation ...

**Comment by mennowitteveen** [comment]
Would like to have a feature like this too. Maybe there is a way to make it less user facing?

**Comment by yusosov** [comment]
@takluyver
Multiple notebook to one kernel?
"...I think it's something we've considered before and aren't interested in supporting..."

Even in 20 years ago in 1999 this feature was present in "Wolfram Mathematica 3.0". It was absolutely obvious to the Wolfram team that it is not possible to manage anything complex enough in one only notebook.

For now jupyter is "a presentation toy", inconvenient for managing anything larger then a thousand lines of code.

I support "Multiple notebook to one kernel" request, and will have to move on to see how to organize my code without jupyter.

**Comment by alancalvitti** [comment]
yusosov, which version JupyterLab supports this? Can't find in 1.0.2 Can you share link to documentation?

**Comment by yusosov** [comment]
@alancalvitti 
hopefully you've already found it. If not
I am on JupyterLab Version 1.0.2
to attach to an existing Kernel click on
whatever is to the left of the small circle in the upper right corner.
(For me it says Python 3). From there you can choose to which of the existing kernels to attach.
Or you can choose to stay not attached or shut down the kernel also.
Or you can execute another notebook like %run /path/to/anotherNotebook.ipynb from an existing notebook.

**Comment by alancalvitti** [comment]
@yosusov, 

Thanks, I didn’t know about the kernel selector. It is flaky in 1.0.2, as caching doesn’t work well and often the child notebook completely loses connection which seems to disconnect the kernel in the parent upon trying to restablish it (refresh page).   Will upgrade to 1.0.5 at some point. 

As far as %run, that’s an option for some work like when utility notebooks are updated, but often, the compute takes too long or there may be cells in development that can raise errors -( I wish options would be included to inactivate cells based on tags.)

**Comment by yusosov** [comment]
I am on Ubuntu 19.04 and JupyterLab 1.0.2. Never had stability issues with
either Jupyter or JupyterLab except they are both "clumsy" to me. Many
thanks the "geniuses" who decided that a web browser and java script are
good for developing GUI applications. Well I'd better shut up.

**Comment by alancalvitti** [comment]
Thanks Yosusov, 

Turns out Safari seems to be the problem - I switched to Chrome which cured most of the issues so far, (Safari couldn’t event let you key-interrupt)

Al
